### PR TITLE
processor ownership

### PR DESCRIPTION
**Pull request by petersteneteg**
This PR changes 2 main thing
1 Processors handled using shared ptrs in the processor network. The reason is 2 fold, better interaction with python processors, and easier handling of background processing. Processor now inherits form std::enabled_shared_from_this. So one can ask for a shared point form this within the processor it self, which is very useful when sending of background jobs from a processor.

2 Python is update to use shared putters for processors and to use a fairly new pybind11 branch by google calls SMART_HOLDERS, basically it has much better support for passing various smart pointers between c++ and python. I also handles trampoline lifetimes. Which mean we no longer just leak all python processors. This change also means that we can remove a lot of other hacks from the bindings.

While at it I also added a VolumePy representation and VolumePy to VolumeGL converters etc, so one can upload data directly from Numpy to OpenGL. Also update a lot of bindings while going through the code.

### CODE REVIEW

File: include/inviwo/core/datastructures/data.h [hunk 1]
```diff
@@ -216,37 +214,40 @@ const T* Data<Self, Repr>::getRepresentation() const {
...
+                        converter->update(srcRepr, destRepr);
+                        lock.lock();
+                    }
+                    lastValidRepresentation_ = destRepr;
+                    lastValidRepresentation_->setValid(true);
+                } else {  // No representation found, create it
+                    lock.unlock();
+                    auto result = converter->createFrom(srcRepr);
+                    if (!result)
+                        throw ConverterException("Converter failed to create", IVW_CONTEXT);
```
Comment (martinfalk) [inline include/inviwo/core/datastructures/data.h]: Failed to create what?

File: include/inviwo/core/processors/processorutils.h [hunk 2]
```diff
@@ -162,6 +164,16 @@ T& trySetProperty(Processor* proc, std::string_view identifier, V&& val, bool re
                         IVW_CONTEXT_CUSTOM("util::trySetProperty"));
     }
 }
+
+/**
+ * @brief Find the module identifier of a registered processor
+ * @param classIdentifier the class identifier of the processor to look for
+ * @param app the InviwoApplication needed to get the modules
+ * @return the identifer of the module that registered the processor or nullopt if not found
```
Comment (martinfalk) [inline include/inviwo/core/processors/processorutils.h]: nullopt not part of std::string_view

File: include/inviwo/core/util/factory.h [hunk 3]
```diff
@@ -64,55 +65,110 @@ class FactoryObservable : public Observable<FactoryObserver<T>> {
...
+    bool hasKey(K key) const;
+    std::vector<Key> getKeys() const;
 
 protected:
-    Map map_;
+    std::map<Key, M*, std::less<>> map_;
+};
+
+/**
+ * T Models the object created
```
Comment (martinfalk) [inline include/inviwo/core/util/factory.h]: consider using `@tparam T` for template parameters.

Comment (martinfalk) [reply include/inviwo/core/util/factory.h]: https://www.doxygen.nl/manual/commands.html#cmdtparam

File: include/inviwo/core/util/factory.h [hunk 4]
```diff
@@ -64,55 +65,110 @@ class FactoryObservable : public Observable<FactoryObserver<T>> {
...
+class SharedStandardFactory : public SharedFactory<T, K, Args...>,
+                              public FactoryRegistar<M, Key, K> {
+public:
+    SharedStandardFactory() = default;
+    virtual std::shared_ptr<T> create(K key, Args... args) const override;
+    virtual bool hasKey(K key) const override { return FactoryRegistar<M, Key, K>::hasKey(key); }
+};
+
+/**
+ * T Models the object created
```
Comment (martinfalk) [inline include/inviwo/core/util/factory.h]: `@tparam`

File: modules/python3/CMakeLists.txt [hunk 5]
```diff
@@ -81,6 +85,8 @@ target_link_libraries(inviwo-module-python3 PUBLIC
 target_include_directories(inviwo-module-python3 PUBLIC
     ${IVW_EXTENSIONS_DIR}/pybind11/include
 )
+target_compile_definitions(inviwo-module-python3 PUBLIC PYBIND11_USE_SMART_HOLDER_AS_DEFAULT)
```
Comment (martinfalk) [inline modules/python3/CMakeLists.txt:85]: is that the reason why you could remove the std::shared_ptr for the exposed data structures?

File: modules/python3/bindings/src/pyimage.cpp [hunk 6]
```diff
@@ -53,8 +56,6 @@
 
 #include <fmt/format.h>
 
-PYBIND11_MAKE_OPAQUE(inviwo::SwizzleMask)
```
Comment (martinfalk) [inline modules/python3/bindings/src/pyimage.cpp:56]: no longer needed?

File: modules/python3/bindings/src/pyvolume.cpp [hunk 7]
```diff
@@ -34,8 +34,10 @@
 #include <inviwopy/inviwopy.h>
 #include <inviwopy/pynetwork.h>
 #include <inviwopy/pyglmtypes.h>
+#include <inviwopy/pyimage.h>  // for the opaque swizzlemask
```
Comment (martinfalk) [inline modules/python3/bindings/src/pyvolume.cpp]: still opaque? That was removed from pyimage.h.

File: modules/python3gl/readme.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Python3GL Module
+
+Description of the Python3GL module
```
Comment (martinfalk) [inline modules/python3gl/readme.md:3]: fixme

File: modules/python3gl/src/volumepytoglconverter.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/python3gl/volumepytoglconverter.h>
+
+namespace inviwo {
+
+std::shared_ptr<VolumeGL> VolumePy2GLConverter::createFrom(
+    std::shared_ptr<const VolumePy> volumeSrc) const {
```
Comment (martinfalk) [inline modules/python3gl/src/volumepytoglconverter.cpp:35]: can we ensure that the correct context is active?

File: modules/python3gl/src/volumepytoglconverter.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+
+    return volumeDst;
+}
+
+void VolumePy2GLConverter::update(std::shared_ptr<const VolumePy> volumeSrc,
+                                  std::shared_ptr<VolumeGL> volumeDst) const {
+
```
Comment (martinfalk) [inline modules/python3gl/src/volumepytoglconverter.cpp]: see comment above re context

File: modules/python3gl/src/volumepytoglconverter.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        auto src = volumeSrc->data().data(0);
+        volumeDst->getTexture()->upload(src);
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+}
+
+std::shared_ptr<VolumePy> VolumeGL2PyConverter::createFrom(
+    std::shared_ptr<const VolumeGL> volumeSrc) const {
+    pybind11::gil_scoped_acquire acquire;
```
Comment (martinfalk) [inline modules/python3gl/src/volumepytoglconverter.cpp]: context

File: modules/python3gl/src/volumepytoglconverter.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    } else {
+        throw Exception("Unable to convert from VolumeGL to VolumePy", IVW_CONTEXT);
+    }
+
+    return volumeDst;
+}
+
+void VolumeGL2PyConverter::update(std::shared_ptr<const VolumeGL> volumeSrc,
+                                  std::shared_ptr<VolumePy> volumeDst) const {
+    pybind11::gil_scoped_acquire acquire;
```
Comment (martinfalk) [inline modules/python3gl/src/volumepytoglconverter.cpp]: context

File: src/qt/editor/processormimedata.cpp [hunk 13]
```diff
@@ -40,13 +40,10 @@
...
-    rendercontext::activateDefault();
-    delete p;
-}
-
-ProcessorMimeData::ProcessorMimeData(std::unique_ptr<Processor> processor)
-    : QMimeData{}, processor_{std::move(processor)} {
+ProcessorMimeData::ProcessorMimeData(std::shared_ptr<Processor> processor)
+    : QMimeData{}
+    , processor_{std::move(processor)}
+    , activateContext_{[]() { rendercontext::activateDefault(); }} {
```
Comment (martinfalk) [inline src/qt/editor/processormimedata.cpp:46]: is `activateContext_` used?

**Review by martinfalk — COMMENTED** [review]
Quite a substantial refactoring.

File: modules/python3/CMakeLists.txt (hunk 5 above)
Comment (petersteneteg) [reply modules/python3/CMakeLists.txt:85]: yes

File: modules/python3/bindings/src/pyimage.cpp (hunk 6 above)
Comment (petersteneteg) [reply modules/python3/bindings/src/pyimage.cpp:56]: Think I moved it

File: include/inviwo/core/datastructures/data.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/data.h]: not part of PR anymore

File: include/inviwo/core/util/factory.h (hunk 4 above)
Comment (petersteneteg) [reply include/inviwo/core/util/factory.h]: fixed in other PR

File: include/inviwo/core/processors/processorutils.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/processors/processorutils.h]: not part of PR anymore

File: include/inviwo/core/util/factory.h (hunk 3 above)
Comment (petersteneteg) [reply include/inviwo/core/util/factory.h]: not part of PR anymore

File: modules/python3/bindings/include/inviwopy/pyproperties.h [hunk 14]
```diff
@@ -59,44 +57,44 @@ using namespace inviwo;
...
+            type = nullptr;
+            return dynamic_cast<const void*>(prop);
         }
-    }
-};
-}  // namespace detail
-}  // namespace pybind11
 
-namespace inviwo {
+        // check if the exact type is resisted in python, then return that.
```
Comment (martinfalk) [inline modules/python3/bindings/include/inviwopy/pyproperties.h:71]:
```suggestion
        // check if the exact type is registered in python, then return that.
```

File: include/inviwo/core/network/processornetwork.h [hunk 15]
```diff
@@ -101,16 +113,9 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
      * @param[in] processor The Processor to be removed.
      * @see addProcessor()
      */
-    void removeProcessor(Processor* processor);
+    std::shared_ptr<Processor> removeProcessor(Processor* processor);
```
Comment (JochenJankowai) [inline include/inviwo/core/network/processornetwork.h:116]: Might want to reflect the return type in the doc

File: src/qt/editor/processormimedata.cpp (hunk 13 above)
Comment (petersteneteg) [reply src/qt/editor/processormimedata.cpp:46]: To activate the right context before deleting the processor.

Comment (petersteneteg) [reply src/qt/editor/processormimedata.cpp:46]: added a comment

File: modules/python3/bindings/src/pyimage.cpp (hunk 6 above)
Comment (petersteneteg) [reply modules/python3/bindings/src/pyimage.cpp:56]: still there

File: modules/python3/bindings/src/pyvolume.cpp (hunk 7 above)
Comment (petersteneteg) [reply modules/python3/bindings/src/pyvolume.cpp]: still there

File: modules/python3gl/src/volumepytoglconverter.cpp (hunk 9 above)
Comment (petersteneteg) [reply modules/python3gl/src/volumepytoglconverter.cpp:35]: This should happen during evaluation, then the right context should be active.
That is the approach we have used so far at least. Changing that would be a different PR.

**Pull request closed without merging**